enhancements/machine-health-checking.md

```diff
@@ -0,0 +1,204 @@
+- [ ] User-facing documentation is created in [openshift/docs
+
+
+## Summary
+
+Enable opt in for automated health checking and remediation of unhealhty nodes backed by machines.
+
+## Motivation
+
+Reduce administrative overhead to run a cluster and ability to respond to failures of machines.
+
+### Goals
+
+- Providing a default remediation strategy for public clouds - Machine deletion.
+
+- Allowing to define different unhealthy criterias for different pools of machines.
+
+- Allowing new and custom remediation strategies developed outside the machine-healthcheck controller, e.g baremetal reboot.
+
+- Providing administrator-defined short-circuiting of automated remediation.
+
+
+### Non-Goals
+
+- Coordinating individual remediation systems. This must be driven by each system however they see fit, e.g. by annotating affected nodes with state and transition time.
+
+- Recording long-term stable history of all health-check failures or remediations.
+
+## Proposal
+
+- As a customer I only care about my apps availability so I want my cluster infrastructure to be self healing and the nodes to be remediated automatically.
+
+- As a cluster admin I want to define my unhealthy criteria as a list of node conditions https://github.com/kubernetes/node-problem-detector.
+
+- As a cluster admin I want to define different unhealthy criteria for pools of machines targeted for different workloads.
+
+- As a cluster admin I want to define different remediation strategies for different cloud environments. E.g deletion or reboot.
+
+- As a cluster admin I want to skip remediation to optionally prevent automated remediation if some threshold of nodes are unhealthy, because this situation may require more careful intervention. E.g a large number of nodes in a single zone are down because it's most likely a networking issue.
+
+Machine health checking is an integration point between node problem detection tooling and remediation strategies.
+
+We will provide a default remediation strategy - delete. It will delete a machine when the backed node meets the unhealthy criteria defined by the user. The upper level machine controller, e.g machineSet will then reconcile towards satisfying the expected number of replicas and will create a new one.
+
+### Implementation Details
+
+#### MachineHealthCheck CRD:
+- Enable watching a pool of machines (based on a label selector).
+- Enable defining an unhealthy criteria (based on a list of node conditions).
+- Enable choosing a remediation strategy.
+- Enable setting a threshold for the number of unhealthy nodes. If the current number is at or above this threshold no further remediation 
+
+
+E.g I want my worker machines in us-east-1a to be remediated when the backed node has `ready=false` condition for more than 10m.
+
+```yaml
+apiVersion: "healthchecking.openshift.io/v1beta1"
+kind: "MachineHealthCheck"
+metadata:
+  name: workers-us-east-1a
+  namespace: openshift-machine-api
+spec:
+  selector:
```

Comment:
This is not a good way to do this.  First, I don't think we should have arbitrary groups like this.  Second, at least use a list of machinesets.  It's not a good strategy to have to duplicate cloud-specific information here, even if it's just the name of a zone.

Reply:
I don't follow what you mean here. This is just an example, as a user you can choose your labels as they better fit their needs.

Reply:
That was accidental, sorry.

I was going to add I would potentially prefer machine selectors over any other mechanism.  Do we have a label that matches the machine set name?  Or does someone have to set that?

Reply:
yeh, machines have the owning machineSet set as a label so you can choose to rely on it, e.g `machine.openshift.io/cluster-api-machineset=agl2309189-ld9sn-worker-us-east-1a` for a MHC 1 - 1 machineSet map